### PR TITLE
Non Existing Directories and Non Existing Artifacts respond with 404

### PR DESCRIPTION
**Pull request by sanket407**
Implemented fix for #418, #419, #420.

Relates to #464.

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 1]
```diff
@@ -135,12 +135,16 @@ public final ArtifactOutputStream getOutputStream(RepositoryPath path) throws IO
     
     @Override
     public URI resolveResource(Repository repository,
-                               String path) throws IOException
+                               String path) 
+                throws IOException,
+                       IllegalArgumentException
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: I don't think any change to this file is necessary. Hence, additional `throws` declaration of `IllegalArgumentException` here is also unnecessary. Would you mind reverting this file to the previous state so there won't be any noise in the Pull Request related with this file ?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java [hunk 2]
```diff
@@ -77,7 +77,17 @@ public InputStream getInputStream(final String storageId,
...
-            final URI resource = layoutProvider.resolveResource(repository, path);
+
+            URI resource;
+            try
+            {
+                resource = layoutProvider.resolveResource(repository, path);
+            }
+            catch(IllegalArgumentException e)
+            {
+                getLogger().debug("Cannot resolve invalid path " + path);
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java]: Could you please add caught Exception `e` as a last argument of `debug` method ? This way we will not loose the stacktrace.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java [hunk 3]
```diff
@@ -77,7 +77,17 @@ public InputStream getInputStream(final String storageId,
...
         {
             final LayoutProvider layoutProvider = layoutProviderRegistry.getProvider(repository.getLayout());
-            final URI resource = layoutProvider.resolveResource(repository, path);
+
+            URI resource;
+            try
+            {
+                resource = layoutProvider.resolveResource(repository, path);
+            }
+            catch(IllegalArgumentException e)
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java]: `catch(IllegalArgumentException e)`

Could you please always format the code fragments you add by the code formatter https://github.com/strongbox/strongbox/wiki/Coding-Convention#code-styles ?

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java [hunk 4]
```diff
@@ -236,11 +234,11 @@ private boolean probeForDirectoryListing(Repository repository,
                                              String path)
     {
         String filePath = path.replaceAll("/", Matcher.quoteReplacement(File.separator));
-
+
```
Comment (fuss86) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/maven/MavenArtifactController.java]: Could you please revert this file to the previous state ? All your changes are related with spaces corrections, so they just bring a noise.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java [hunk 5]
```diff
@@ -77,7 +77,17 @@ public InputStream getInputStream(final String storageId,
...
-            final URI resource = layoutProvider.resolveResource(repository, path);
+
+            URI resource;
+            try
+            {
+                resource = layoutProvider.resolveResource(repository, path);
+            }
+            catch (IllegalArgumentException e)
+            {
+                getLogger().debug("Cannot resolve invalid path " + path + ". Exception : " + e);
```
Comment (fuss86) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/proxied/ProxyRepositoryArtifactResolver.java]: `getLogger().debug("Cannot resolve invalid path " + path + ". Exception : " + e);`

I meant

`getLogger().debug("Cannot resolve invalid path " + path, e);`

File: strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java [hunk 6]
```diff
@@ -472,7 +473,31 @@ public void testDeleteArtifactDirectory()
         assertFalse("Failed to delete artifact file '" + deletedArtifact.getAbsolutePath() + "'!",
                     deletedArtifact.exists());
     }
-
+    
+    @Test
+    public void testNonExistingDirectoryDownload()
+    {
+        String path = "/storages/storage-common-proxies/maven-central/john/doe/";
```
Comment (fuss86) [inline strongbox-web-core/src/test/java/org/carlspring/strongbox/controllers/maven/MavenArtifactControllerTest.java]: `String path = "/storages/storage-common-proxies/maven-central/john/doe/";`

Could you please write one more test for:

`String path = "/storages/storage-common-proxies/maven-central/john/doe";` ? Notice removed trailing slash. Thanks

**Comment by fuss86** [comment]
Pull looks fine by me :) Thanks @sanket407 !

**Comment by steve-todorov** [comment]
Hi @sanket407 and thanks for your PR! Looks great but you should not print the `java.lang.IllegalArgumentException: Illegal character` as @fuss86 has suggested in his comment earlier (https://github.com/strongbox/strongbox/pull/463#discussion_r159052950). You should only print a message because we already have too much logging info:

```
getLogger().debug("Cannot resolve invalid path " + path);
```

**Comment by carlspring** [comment]
I'd actually say -- print nothing at all. We have tons and tons of debugging output. While developing the code, it's fine to have a line there that prints it out, but, to be honest, there will be many cases where people try to resolve an artifact and it's simply not there. These are all perfectly legal cases. If we keep printing these exceptions, or even one-liners all the time, the logs will eventually produce a lot of unnecessary noise...

**Comment by sanket407** [comment]
@carlspring  ok ! I will just keep a comment in its place ! Just for further reference !

**Comment by carlspring** [comment]
Okay, that sounds good to me.
Let me know when it could be merged. :)

**Comment by sanket407** [comment]
Its done !

**Comment by carlspring** [comment]
Thanks, @sanket407 !

**Pull request merged**